base/suffix_array.cpp

```diff
@@ -0,0 +1,252 @@
+#include "base/suffix_array.hpp"
+
+#include "base/assert.hpp"
+
+#include <limits>
+
+using namespace std;
+
+namespace
+{
+bool LEQ(size_t a1, size_t a2, size_t b1, size_t b2)
+{
+  if (a1 != b1)
+    return a1 < b1;
+  return a2 <= b2;
+}
+
+bool LEQ(size_t a1, size_t a2, size_t a3, size_t b1, size_t b2, size_t b3)
+{
+  if (a1 != b1)
+    return a1 < b1;
+  return LEQ(a2, a3, b2, b3);
+}
+
+template <typename Values>
+void RadixSort(size_t n, size_t const * keys, size_t maxValue, Values const & values,
+               size_t * resultKeys)
+{
+  std::vector<size_t> count(maxValue);
```

Comment:
Sounds good, but I'd prefer to rename this to numValues, and get rid of headache here that maxValue may be numeric_limits<size_t>::max(), what do you think?